src/Arcus.WebApi.Logging/RequestTrackingMiddleware.cs

```diff
@@ -0,0 +1,125 @@
+            _logger = logger;
+        }
+
+        /// <summary>
+        /// Logs every incoming HTTP request.
+        /// </summary>
+        /// <param name="httpContext">The current HTTP context.</param>
+        public async Task Invoke(HttpContext httpContext)
+        {
+            var stopwatch = Stopwatch.StartNew();
+            if (_options.IncludeRequestBody)
+            {
+                httpContext.Request.EnableBuffering();
+            }
+
+            try
+            {
+                await _next(httpContext);
+            }
+            finally
+            {
+                stopwatch.Stop();
+                await TrackRequestAsync(httpContext, stopwatch.Elapsed);
+            }
+        }
+
+        private async Task TrackRequestAsync(HttpContext httpContext, TimeSpan duration)
+        {
+            try
+            {
+                IDictionary<string, StringValues> headers = 
+                    _options.IncludeRequestHeaders
+                        ? SanitizeRequestHeaders(httpContext.Request.Headers) ?? new Dictionary<string, StringValues>()
+                        : new Dictionary<string, StringValues>();
+                
+                
+                IDictionary<string, StringValues> body = 
+                    _options.IncludeRequestBody
+                        ? await SanitizeRequestBodyAsync(httpContext.Request.Body) ?? new Dictionary<string, StringValues>()
+                        : new Dictionary<string, StringValues>();
+                
+                Dictionary<string, object> logContext = headers.Concat(body).ToDictionary(kv => kv.Key, kv => (object) kv.Value);
+                _logger.LogRequest(httpContext.Request, httpContext.Response, duration, logContext);
+            }
+            catch (Exception exception)
+            {
+                _logger.LogCritical(exception, "Failed to track request");
+            }
+        }
+
+        /// <summary>
+        /// Extracts information from the given HTTP <paramref name="requestHeaders"/> to include in the request tracking context.
+        /// </summary>
+        /// <param name="requestHeaders">The headers of the current HTTP request.</param>
+        protected virtual IDictionary<string, StringValues> SanitizeRequestHeaders(IDictionary<string, StringValues> requestHeaders)
+        {
+            return requestHeaders.Where(header => !_options.OmittedHeaderNames.Contains(header.Key));
+        }
+
+        private static async Task<IDictionary<string, StringValues>> SanitizeRequestBodyAsync(Stream requestStream)
+        {
+            if (!requestStream.CanRead)
+            {
+                return new Dictionary<string, StringValues>();
+            }
+
+            if (requestStream.CanSeek)
+            {
+                if (requestStream.Position > 0)
+                {
+                    requestStream.Position = 0;
+                }
+            }
+
+            using (var reader = new StreamReader(requestStream))
```

Comment:
Can we dispose this? What happens if controller reads it again? Is that what the buffering option covers?

Reply:
Yes, the reading inside the controller comes first, actually, (during the `await _next(httpContext)` call), but indeed after that we have to read it again here. That's indeed where the buffering comes into play.

Reply:
Except if there's another middleware component after this one that also needs to read the request body, I don't see a problem 🤔 . It may get disposed anyway after this request middleware, but you can argue if it's the middleware's job to do so.

Reply:
Perfect, we're good to go then!